### PR TITLE
wip: gateway implementation

### PR DESCRIPTION
**Pull request by lahgolz**

### CODE REVIEW

File: src/constants.ts [hunk 1]
```diff
@@ -3,3 +3,11 @@ const { homepage, version } = require('../package') as Record<string, string>;
 
 export { version };
 export const USER_AGENT = `DiscordBot (${homepage.split('#')[0]}, ${version})`;
+
+export const WSEvents = {
```
Comment (Aakodal) [inline src/constants.ts]: (Screaming) snake case should be used here
```suggestion
export const WS_EVENTS = {
```

File: src/gateway/WebSocketClient.ts [hunk 2]
```diff
@@ -0,0 +1,139 @@
+import { EventEmitter } from 'events';
+import { Inflate } from 'zlib-sync';
+import ws from 'ws';
+import * as dgateway from 'discord-api-types/gateway/v8';
+import { Client } from '../clients/Client';
+import { WSEvents } from '../constants';
```
Comment (Aakodal) [inline src/gateway/WebSocketClient.ts]:
```suggestion
import { WS_EVENTS } from '../constants';
```

File: src/gateway/WebSocketClient.ts [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+		if (this.connection?.readyState === ws.OPEN && this.status === Status.READY) {
+			return;
+		}
+
+		return new Promise((resolve, reject) => {
+			this.once(WSEvents.READY, () => resolve());
+			this.once(WSEvents.RESUMED, () => resolve());
+			this.once(WSEvents.CLOSE, (event: ws.CloseEvent) => reject(event));
+			this.once(WSEvents.INVALID_SESSION, () => reject());
+			this.once(WSEvents.DESTROYED, () => reject());
```
Comment (Aakodal) [inline src/gateway/WebSocketClient.ts]:
```suggestion
			this.once(WS_EVENTS.READY, () => resolve());
			this.once(WS_EVENTS.RESUMED, () => resolve());
			this.once(WS_EVENTS.CLOSE, (event: ws.CloseEvent) => reject(event));
			this.once(WS_EVENTS.INVALID_SESSION, () => reject());
			this.once(WS_EVENTS.DESTROYED, () => reject());
```

File: src/gateway/WebSocketClient.ts (hunk 2 above)
Comment (Mesteery) [reply src/gateway/WebSocketClient.ts]: In this case : `WEBSOCKET_EVENTS` 😄

File: src/gateway/WebSocketClient.ts (hunk 3 above)
Comment (Mesteery) [reply src/gateway/WebSocketClient.ts]: Up

File: src/constants.ts (hunk 1 above)
Comment (Mesteery) [reply src/constants.ts]: Down

File: src/gateway/WebSocketClient.ts (hunk 2 above)
Comment (Mesteery) [reply src/gateway/WebSocketClient.ts]: (Also, this should be in the `gateway` folder, not in constants.ts)

File: src/clients/BaseClient.ts [hunk 4]
```diff
@@ -50,9 +50,14 @@ export interface BaseClientOptions {
 	ws?: {
 		/**
 		 * WebSocket protocol version
-		 * @default 13
+		 * @default 8
```
Comment (Mesteery) [inline src/clients/BaseClient.ts]:
```suggestion
		 * @default 13
```

File: src/clients/BaseClient.ts [hunk 5]
```diff
@@ -50,9 +50,14 @@ export interface BaseClientOptions {
...
 		 */
 		protocolVersion?: 8 | 13;
+		/**
+		 * Us of zlib compression/decompression
```
Comment (Mesteery) [inline src/clients/BaseClient.ts]:
```suggestion
		 * Use of zlib compression/decompression
```

File: src/clients/BaseClient.ts [hunk 6]
```diff
@@ -50,9 +50,14 @@ export interface BaseClientOptions {
 	ws?: {
```
Comment (Mesteery) [inline src/clients/BaseClient.ts]: This should be in `Client`, not in `BaseClient`

File: src/clients/BaseClient.ts [hunk 7]
```diff
@@ -66,7 +71,8 @@ const defaultOptions: DeepRequired<BaseClientOptions> = {
 		api: 'https://discord.com/api/v8',
 	},
 	ws: {
```
Comment (Mesteery) [inline src/clients/BaseClient.ts]: Same as above

File: src/clients/Client.ts [hunk 8]
```diff
@@ -14,18 +15,33 @@ const defaultOptions: DeepRequired<ClientOptions> = {
 		api: 'https://discord.com/api/v8',
 	},
 	ws: {
-		protocolVersion: 13,
+		protocolVersion: 8,
```
Comment (Mesteery) [inline src/clients/Client.ts]:
```suggestion
		protocolVersion: 13,
```

File: src/gateway/WebSocket.ts [hunk 9]
```diff
@@ -3,17 +3,17 @@ import BaseWebSocket from 'ws';
```
Comment (Mesteery) [inline src/gateway/WebSocket.ts]: This file is useless.

File: src/gateway/WebSocketClient.ts [hunk 10]
```diff
@@ -1,34 +1,60 @@
...
-	DISCONNECTED,
-	WAITING_FOR_GUILDS,
-	IDENTIFYING,
-	RESUMING,
+import { GatewayURLQuery, WebSocket } from './WebSocket';
+import { pack, unpack } from '.';
+import { Status, WebSocketManager } from './WebSocketManager';
+import { constants, createInflate, Inflate } from 'zlib';
+
+export const enum WEBSOCKET_EVENTS {
```
Comment (Mesteery) [inline src/gateway/WebSocketClient.ts]:
```suggestion
export const enum WebSocketEvents {
```

File: src/gateway/WebSocketClient.ts [hunk 11]
```diff
@@ -37,103 +63,407 @@ export class WebSocketClient extends EventEmitter {
...
+		if (this.readyTimeout) {
+			this.client.clearTimeout(this.readyTimeout);
+
+			this.readyTimeout = undefined;
+		}
+
+		if (!this.expectedGuilds?.size) {
+			this.status = Status.READY;
+
+			// this.emit(ShardEvents.ALL_READY);
```
Comment (Mesteery) [inline src/gateway/WebSocketClient.ts:263]: ?

File: src/gateway/WebSocketClient.ts [hunk 12]
```diff
@@ -37,103 +63,407 @@ export class WebSocketClient extends EventEmitter {
...
+
 			return;
 		}
 
-		// this.onPacket(packet);
+		this.readyTimeout = this.client.setTimeout(() => {
+			this.readyTimeout = undefined;
+			this.status = Status.READY;
+
+			// this.emit(ShardEvents.ALL_READY, this.expectedGuilds);
```
Comment (Mesteery) [inline src/gateway/WebSocketClient.ts:271]: ?

File: src/gateway/WebSocketClient.ts [hunk 13]
```diff
@@ -37,103 +63,407 @@ export class WebSocketClient extends EventEmitter {
...
+		this.send({ op: dgateway.GatewayOPCodes.Heartbeat, d: this.sequence }, true);
+	}
+
+	private ackHeartbeat(): void {
+		this.lastHeartbeatAcked = true;
+		this.ping = Date.now() - this.lastPingTimestamp;
+	}
+
+	private identify(): void {
+		return this.sessionID ? this.identifyResume() : this.identifyNew();
```
Comment (Mesteery) [inline src/gateway/WebSocketClient.ts]:
```suggestion
		return this.sessionId ? this.identifyResume() : this.identifyNew();
```

File: src/gateway/WebSocketClient.ts [hunk 14]
```diff
@@ -37,103 +63,407 @@ export class WebSocketClient extends EventEmitter {
...
+			this.ratelimit.queue.push(data);
 		} else {
-			this.inflate?.push(transformedData, isFlush && zlibSync.Z_SYNC_FLUSH);
+			this.ratelimit.queue.unshift(data);
+		}
+
+		this.processQueue();
+	}
+
+	private _send(data: any) {
```
Comment (Mesteery) [inline src/gateway/WebSocketClient.ts]: $send ?

File: src/gateway/WebSocketClient.ts [hunk 15]
```diff
@@ -37,103 +63,407 @@ export class WebSocketClient extends EventEmitter {
...
+		if (!this.connection || this.connection.readyState !== WebSocket.OPEN) {
+			this.destroy({ closeCode: 4000 });
+
+			return;
 		}
 
-		return isFlush ? this.inflate?.result : undefined;
+		this.connection.send(pack(data), () => {
+			/*if (err) {
+				this.client.emit(Events.SHARD_ERROR, err, this.id);
```
Comment (Mesteery) [inline src/gateway/WebSocketClient.ts]: ?

File: src/gateway/WebSocketClient.ts [hunk 16]
```diff
@@ -37,103 +63,407 @@ export class WebSocketClient extends EventEmitter {
...
+			return;
+		}
+
+		this.connection.onopen = this.connection.onclose = this.connection.onerror = this.connection.onmessage = () => {};
 	}
 
-	private onClose(): void {
-		return;
+	private _emitDestroyed(): void {
+		// this.emit(ShardEvents.DESTROYED);
```
Comment (Mesteery) [inline src/gateway/WebSocketClient.ts:466]: ?

File: src/gateway/WebSocketManager.ts [hunk 17]
```diff
@@ -0,0 +1,198 @@
...
+	public constructor(public readonly client: Client) {
+		super();
+	}
+
+	get ping() {
+		return this.webSocketClient?.ping;
+	}
+
+	public async connect(): Promise<void> {
+		/*const invalidToken = new Exception(WEBSOCKET_CODES[GatewayCloseCodes.AuthenticationFailed]);
```
Comment (Mesteery) [inline src/gateway/WebSocketManager.ts:54]: ?

File: src/utils/sleep.ts [hunk 18]
```diff
@@ -0,0 +1,3 @@
+export function sleep(milliseconds: number) {
+	return new Promise((resolve) => setTimeout(resolve, milliseconds));
```
Comment (Mesteery) [inline src/utils/sleep.ts]: promisify(setTimeout)

File: src/gateway/WebSocketManager.ts [hunk 19]
```diff
@@ -0,0 +1,198 @@
...
+const UNRECOVERABLE_CLOSE_CODES = [4004];
+const UNRESUMABLE_CLOSE_CODES = [1000];
+
+export class WebSocketManager extends EventEmitter {
+	private webSocketClient?: WebSocketClient;
+	private packetQueue: (dgateway.GatewaySendPayload | dgateway.GatewayReceivePayload | dgateway.GatewayDispatchPayload)[] = [];
+	private status = Status.IDLE;
+	private destroyed = false;
+	private reconnecting = false;
+	public gateway = 'wss://gateway.discord.gg/';
```
Comment (Mesteery) [inline src/gateway/WebSocketManager.ts:43]: Move this to ws options

File: src/gateway/WebSocketClient.ts [hunk 20]
```diff
@@ -37,103 +63,407 @@ export class WebSocketClient extends EventEmitter {
...
 
-		if (!zlibSync) {
-			return transformedData;
+		this.status = Status.IDENTIFYING;
+
+		this.send(
+			{
+				op: dgateway.GatewayOPCodes.Identify,
+				d: {
+					...this.client.options.ws,
```
Comment (Mesteery) [inline src/gateway/WebSocketClient.ts]: options.ws can contains others options that should not be sent

File: src/gateway/WebSocketManager.ts [hunk 21]
```diff
@@ -0,0 +1,198 @@
...
+			return;
+		}
+
+		this.reconnecting = true;
+
+		try {
+			await this.createClient();
+		} catch (error) {
+			if (error.httpStatus !== 401) {
+				await sleep(5000);
```
Comment (Mesteery) [inline src/gateway/WebSocketManager.ts]: Why not client.setTimeout ?

Comment (Mesteery) [reply src/gateway/WebSocketManager.ts]: (And why wait 5000ms ?)

**Review by Mesteery — CHANGES_REQUESTED** [review]
There are a lot of comments that should be removed

**Comment by lahgolz** [comment]
PR closed while we work on it.

**Pull request closed without merging**